AI #1785: Add Invoice Detail and Billing Period datasets (#1900)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/datasets/invoice_detail/columns/invoicedetailgrain.md

# Invoice Detail Grain

> Invoice Detail Grain represents the set of key-value pairs that defines the granularity of the invoice line item. This data is represented as key-value pairs instead of separate columns because the grain may vary from one record to the next, both on a single invoice and across invoice issuers; this enables FinOps practitioners to have a single point of reference for the known universe of [Invoice Detail](#datasets.invoicedetail) granularities for use in any necessary downstream data transformations. This information helps FinOps practitioners understand the specific grouping or aggregation level of a given [Invoice Detail](#datasets.invoicedetail) record, such as whether it represents a summary by service, resource, or another dimensional attribute.

Reviewer: Feel free to discard this, because I'm not 100% sure what to cut/replace exactly, but this definition feels dense and I wonder if practitioners are going to follow the points being made here. I wonder if something simpler like
"Invoice Detail Grain captures the dimensional attributes that define how a line item is grouped on an invoice, expressed as flexible key-value pairs." may be a better intro?

Author: Happy to incorporate a revision if you make a suggestion.

Author: Resolving in favor of [this thread](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1900/files#r2815135171).

## Requirements

InvoiceDetailGrain adheres to the following requirements:

* InvoiceDetailGrain MUST be of type JSON.
* InvoiceDetailGrain MUST conform to [KeyValueFormat](#attributes.key-valueformat) requirements.
* InvoiceDetailGrain MUST NOT be null when one or more properties uniquely define the granularity of the invoice line item.
* InvoiceDetailGrain MUST contain the set of all properties that uniquely define the granularity of the invoice line item.
* InvoiceDetailGrain SHOULD use the applicable FOCUS-defined Invoice Detail Grain properties listed below to represent the granularity of the invoice line item.
* InvoiceDetailGrain MUST include all custom Invoice Detail Grain properties that are applicable to the granularity of the invoice line item when there is no equivalent FOCUS-defined property.
* InvoiceDetailGrain property keys SHOULD conform to [PascalCase](#glossary:pascalcase) format.

> * InvoiceDetailGrain property keys MUST begin with the string "x_" unless it is a FOCUS-defined property.

Reviewer: I wonder if it's worth adding a note clarifying that custom keys are case-sensitive and values must be scalar (string, number, boolean) unless otherwise specified?

Author: Yes, you're right.  More globally, all keys need to PascalCase.  I forgot to lift that requirement from `SkuPriceDetails`, which I was using as a template, given that the columns are pretty similar.

See the third requirement.  This should address that concern of yours.

# Invoice Detail Grain

> Invoice Detail Grain represents the set of key-value pairs that defines the granularity of the invoice line item. This data is represented as key-value pairs instead of separate columns because the grain may vary from one record to the next, both on a single invoice and across invoice issuers; this enables FinOps practitioners to have a single point of reference for the known universe of [Invoice Detail](#datasets.invoicedetail) granularities for use in any necessary downstream data transformations. This information helps FinOps practitioners understand the specific grouping or aggregation level of a given [Invoice Detail](#datasets.invoicedetail) record, such as whether it represents a summary by service, resource, or another dimensional attribute.

Reviewer: Here's my suggestion, breaking these two concepts into 2 distinct paragraphs. Please tweak to your liking:

```suggestion
Invoice Detail Grain represents the set of key-value pairs that defines the granularity of an invoice line item. The grain may vary from one record to the next, both within a single invoice and across invoice issuers.

Key-value pairs are used instead of separate columns to accommodate this variability. This gives FOCUS consumers a single point of reference for all known [Invoice Detail](#datasets.invoicedetail) granularities, supporting downstream data transformations such as reconciliation and cost allocation.
```

Also flagging: the PascalCase requirement appears twice (lines 10 and 15), once without and once with the glossary link. I think the first instance on line 10 can be removed.

Author: Noted.  How about this?

```suggestion
Invoice Detail Grain represents the set of key-value pairs that defines the granularity of an invoice line item. The grain may vary from one record to the next, both within a single invoice and across invoice issuers, and key-value pairs are used instead of separate columns to accommodate this variability. 

This gives FinOps practitioners a single point of reference for all possible [Invoice Detail](#datasets.invoicedetail) granularities (e.g., SKU, service, resource, custom dimension), supporting downstream data transformations such as reconciliation and cost allocation.
```

## Requirements

InvoiceDetailGrain adheres to the following requirements:

* InvoiceDetailGrain MUST be of type JSON.
* InvoiceDetailGrain MUST conform to [KeyValueFormat](#attributes.key-valueformat) requirements.

> * InvoiceDetailGrain property keys SHOULD conform to PascalCase format.

Author: Per @Matt-Cowsert, we should delete this.  We have it five rows down.
```suggestion
```

* InvoiceDetailGrain MUST NOT be null when one or more properties uniquely define the granularity of the invoice line item.
* InvoiceDetailGrain MUST contain the set of all properties that uniquely define the granularity of the invoice line item.
* InvoiceDetailGrain SHOULD use the applicable FOCUS-defined Invoice Detail Grain properties listed below to represent the granularity of the invoice line item.
* InvoiceDetailGrain MUST include all custom Invoice Detail Grain properties that are applicable to the granularity of the invoice line item when there is no equivalent FOCUS-defined property.
* InvoiceDetailGrain property keys SHOULD conform to [PascalCase](#glossary:pascalcase) format.
* InvoiceDetailGrain property keys MUST begin with the string "x_" unless it is a FOCUS-defined property.
* FOCUS-defined InvoiceDetailGrain properties adhere to the following additional requirements:
  * Property key MUST match the spelling and casing specified for the FOCUS-defined property.

>   * Property value MUST be of the type specified for that property.

Reviewer 2: Let's discuss how we can unify the way we are specifying normative text for the contents of the JSON

Author: Happy to discuss.  I lifted this verbiage verbatim from SkuPriceDetails.  

Let's discuss during our weekly sync tomorrow.  Then I can figure out the next step in the working group.

Author: Any further discussion needed here @mike-finopsorg?  As I mentioned previously, I'm following the same pattern as that in SkuPriceDetails.

Reviewer 2: I think KeyValueFormat columns can remain in this structure as it's not defining another object. Will resolve this comment.

# Invoice Detail Grain

> Invoice Detail Grain represents the set of key-value pairs that defines the granularity of an invoice line item. The grain may vary from one record to the next, both within a single invoice and across invoice issuers, and key-value pairs are used instead of separate columns to accommodate this variability.

Reviewer 3: To ease the life of a practitioner, I wonder whether the schema of invoice detail grains could be specified in the meta-data, so that the parsing of invoice detail grains could be configured properly based on the meta-data.

@shawnalpay: this comment should not be blocking. Feel free to resolve it. You can also let me know if it is relevant to create a candidate feature from it for a future FOCUS version.

Author: @marc-perreaut I agree that we should consider this.  We do not currently do that for other similar columns, so I would rather record this need for a future release, which I have now done as a comment on FR #1937 [here](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/issues/1937#issuecomment-3998957156).  Feel free to add more color there.

Resolving.

This gives FinOps practitioners a single point of reference for all possible [Invoice Detail](#datasets.invoicedetail) granularities (e.g., SKU, service, resource, custom dimension), supporting downstream data transformations such as reconciliation and cost allocation.

## Requirements

> InvoiceDetailGrain adheres to the following requirements:

Reviewer 4: ```suggestion
InvoiceDetailGrain MUST adhere to the following requirements:
```

* InvoiceDetailGrain MUST be of type JSON.
* InvoiceDetailGrain MUST conform to [KeyValueFormat](#attributes.key-valueformat) requirements.
* InvoiceDetailGrain MUST NOT be null when one or more properties uniquely define the granularity of the invoice line item.
* InvoiceDetailGrain MUST contain the set of all properties that uniquely define the granularity of the invoice line item.
* InvoiceDetailGrain SHOULD use the applicable FOCUS-defined Invoice Detail Grain properties listed below to represent the granularity of the invoice line item.
* InvoiceDetailGrain MUST include all custom Invoice Detail Grain properties that are applicable to the granularity of the invoice line item when there is no equivalent FOCUS-defined property.
* InvoiceDetailGrain property keys SHOULD conform to [PascalCase](#glossary:pascalcase) format.
* InvoiceDetailGrain property keys MUST begin with the string "x_" unless it is a FOCUS-defined property.

> * FOCUS-defined InvoiceDetailGrain properties adhere to the following additional requirements:

Reviewer 4: ```suggestion
* FOCUS-defined InvoiceDetailGrain properties MUST adhere to the following requirements:
```
